File: python/0027移除元素.py

nums=[1,2,3,4]

# 首尾指针
def RemoveElement1(nums,val):
    if(not nums):
        return 0
    rear=len(nums)-1
    idx=0

    while(idx<=rear):
        if(nums[idx]==val and nums[rear]!=val):
            nums[idx],nums[rear]=nums[rear],nums[idx]

        if(nums[rear]==val):
            rear-=1
        if(nums[idx]!=val):
            idx+=1

    return rear+1

# 双指针（删除的元素会被覆盖）
def RemoveElement2(nums,val):
    res_idx=0
    for i in range(len(nums)):
        if(nums[i]!=val):
            nums[res_idx]=nums[i]
            res_idx+=1
    return res_idx
    
print(RemoveElement2(nums,1))
print(nums)

File: cpp/0027移除元素.cpp
#include<vector>
#include<iostream>
using namespace std;

// 首尾指针
int RemoveElement1(vector<int>&nums,int val)
{
    int front = 0;
    int rear = nums.size()-1;
    int t;
    while (front <= rear)
    {
        if(nums[front]==val&&nums[rear]!=val)
        {
            t = nums[front];
            nums[front] = nums[rear];
            nums[rear] = t;
        }

        if (nums[rear] == val)
            --rear;
        if(nums[front]!=val)
            ++front;
    }
    return rear+1;
}

//双指针
int RemoveElement2(vector<int> &nums, int val)
{
    int res_idx = 0;
    int size = nums.size();

    for (int i = 0;i<size;++i)
    {
        if(nums[i]!=val)
            nums[res_idx++] = nums[i];
    }

    return res_idx;
}

int main()
{
    vector<int> nums = {1};
    int size = RemoveElement1(nums, 1);
    cout << size << endl;
    for (int i = 0; i < size; i++)
        cout << nums[i] << " ";
}